baseplate/server.go

```diff
@@ -0,0 +1,75 @@
+package baseplate
+
+import (
+	"errors"
+	"io"
+	"io/ioutil"
+	"time"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/secrets"
+)
+
+// ServerConfig is a general purpose config for assembling a BaseplateServer
+type ServerConfig struct {
+	Addr string
+
+	Timeout time.Duration
+
+	Log struct {
+		Level log.Level
+	}
+
+	Metrics struct {
+		Namespace string
+		Endpoint  string
+	}
+
+	Secrets struct {
+		Path string
+	}
+
+	Sentry struct {
+		DSN         string
+		Environment string
+		SampleRate  float64
+	}
+
+	Tracing struct {
+		Namespace     string
+		Endpoint      string
+		RecordTimeout time.Duration
+		SampleRate    float64
+	}
+}
+
+// Server is the primary interface for baseplate servers.
+type Server interface {
+	io.Closer
+
+	Config() ServerConfig
+	Secrets() *secrets.Store
+	Serve() error
+}
+
+// ParseServerConfig will populate a ServerConfig from a YAML file.
+func ParseServerConfig(path string) (*ServerConfig, error) {
+	cfg := &ServerConfig{}
+	if path == "" {
+		return nil, errors.New("no config path given")
+	}
+
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = yaml.Unmarshal(data, cfg); err != nil {
```

Comment:
This is minor since this is once on start up thing instead of once per request thing, but yaml package does have [`NewDecoder`](https://pkg.go.dev/gopkg.in/yaml.v2?tab=doc#NewDecoder) to take an `io.Reader` directly, which means there's no need to read the file fully into memory before passing that into yaml for parsing.

Reply:
I still prefer this to be fixed. Although we care less about performance here, I consider baseplate.go codebase an example for engineers less familiar with go to follow, and this is a bad example.